pkg/cableengine/ipsec/ipsec.go

```diff
@@ -95,7 +95,26 @@ func (i *Engine) StartEngine(ignition bool) error {
 		forwardToSubPostroutingRuleSpec := []string{"-j", "SUBMARINER-POSTROUTING"}
 		ipt.AppendUnique("nat", "POSTROUTING", forwardToSubPostroutingRuleSpec...)
 		forwardToSubForwardRuleSpec := []string{"-j", "SUBMARINER-FORWARD"}
-		ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
+		rules, err := ipt.List("filter", "FORWARD")
+		if err != nil {
+			klog.Fatalf("error listing the rules in FORWARD chain: %v", err)
+		}
+
+		insertAt := -1
+		for i, rule := range rules {
+			if rule == "-A FORWARD -j REJECT --reject-with icmp-host-prohibited" {
+				insertAt = i
+				break
+			}
+		}
+
+		// If the FORWARD Chain does not have any REJECT rule, append the
+		// forwardToSubForwardRuleSpec rule, otherwise insert it at the appropriate location.
+		if insertAt == -1 {
+			ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
+		} else if rules[insertAt-1] != "-A FORWARD -j SUBMARINER-FORWARD" {
```

Comment:
I'm afraid, insertAt cannot be 0 (we would get an error when the iptables command is executed). 
Regarding the other point, I hope when you say "FORWARD", you mean "SUBMARINER-FORWARD". So, if there is already a SUBMARINER-FORWARD rule, but there is no REJECT rule, its still not a problem because we use AppendUnique. I verified this scenario.

Reply:
Re: insertAt being 0, I was referring to rules[insertAt - 1] which would blow up with an array index-out-of-bounds.  And yes by "FORWARD", I mean "-A FORWARD -j SUBMARINER-FORWARD".  Ack on the AppendUnique. Than perhaps we can just check:

    else if insertAt == 0 || rules[insertAt - 1] != "..." {
        // insert
    }

Reply:
As mentioned in my earlier comments, if a Chain exists (i.e., in our case, its the FORWARD chain which is one of the standard Chains which would exist) but has no rules, ipt.List(...) returns success and the len(rules) will be 1. The zeroth entry will be the default policy associated with the chain. 

So, even in such situations (i.e., empty FORWARD chain), when the following code is executed, insertAt will not be zero (since 0th entry is the default policy of the Chain and not the rules. The rules start from 1).

                for i, rule := range rules {
                        if rule == "-A FORWARD -j REJECT --reject-with icmp-host-prohibited" {
                                insertAt = i 
                                break
                        }
                }

IMHO, checking "else if insertAt == 0 || ...." is redundant. Also, if we execute ipt.Insert(...) with insertAt as 0, it will return an error.

Reply:
OK - as long as that behavior is documented and guaranteed and across all platforms etc.  I'm just playing devil's advocate here wrt strict coding, ie not assuming anything and guarding against index OOB.  It may be redundant but IMO it doesn't hurt to check (just in case...). Can you at least add a comment explaining the expected behavior of ipt and what we're assuming wrt our index logic?

Reply:
I see your point. Let me update the code. Thanks for reviewing.